### PR TITLE
Images built from scratch w/CircleCI configs

### PR DESCRIPTION
**Pull request by dragonsahead**
Fixes https://github.com/metabase/metabase/issues/18650

EDIT: new images have:
### Build (arm64 and amd64 compatible) between 248Mb and 347Mb
- 1.11.0.1100 Clojure CLI version
- Node LTS (16 by the time this image was built)
- Yarn
- wget
- netcat
- git

### Test (only amd64) between 609Mb and 695Mb. On top of Build:
- Firefox
- Chrome

tests on: https://app.circleci.com/pipelines/github/metabase/metabase?branch=check-new-builder&filter=all

### CODE REVIEW

**Comment by dragonsahead** [comment]
So this one is ready now, as I separated between builders and testers so we can have smaller images that spin up faster

**Comment by dragonsahead** [comment]
So after a long wait, CircleCI people published https://discuss.circleci.com/t/next-gen-clojure-convenience-image-beta/42607

File: .github/workflows/build-and-push.yml
Comment (nemanjaglumac) [inline .github/workflows/build-and-push.yml]:
```suggestion
name: Build and push to Dockerhub

on:
  push:
    branches:
      - master

jobs:
  build-java:
    runs-on: ubuntu-20.04
    timeout-minutes: 30
    strategy:
      matrix:
        version: [8, 11, 17]
    steps:
      - uses: actions/checkout@v2
      - name: Set up QEMU
        uses: docker/setup-qemu-action@v1
      - name: Set up Docker Buildx
        id: buildx
        uses: docker/setup-buildx-action@v1
      - name: Login to Docker Hub
        uses: docker/login-action@v1
        with:
          username: ${{ secrets.DOCKERHUB_USERNAME }}
          password: ${{ secrets.DOCKERHUB_TOKEN }}
      - name: Create the Java ${{ matrix.version }} builder image
        # we still can't add linux/arm64/v8 because of a segfault on yarn installation on 20.04
        run: |
          docker buildx build
          -f java-${{ matrix.version }}/build/Dockerfile
          --platform linux/amd64
          -t metabase/ci:java-${{ matrix.version }}-clj-1.10.1040.12-2021-build --push --compress --no-cache .
      - name: Create the Java ${{ matrix.version }} testing image
        run: |
          docker buildx build
          -f java-${{ matrix.version }}/test/Dockerfile
          -t metabase/ci:java-${{ matrix.version }}-clj-1.10.1040.12-2021-node-browsers --push --compress --no-cache .

```

Nit: why don't we use `strategy` matrix to DRY this code out a little bit?
Nit 2: Suggested giving commands their own line in the `run` step for better readability

Comment (dragonsahead) [reply .github/workflows/build-and-push.yml]: makes lot of sense!

**Review by ariya — APPROVED** [review]

**Comment by jeff303** [comment]
@paoliniluis , waiting on anything else to merge this?

**Review by camsaul — CHANGES_REQUESTED** [review]
Can we put the new Dockerfiles inside a top-level directory? If I see a `java-8` and `java-17` at the top level I'm going to get confused and think there are source files in them, because we already have a top-level `java` directory with our one `.java` source file. If I look in them and see Docker files I'm maybe going to mistakenly think those are the Dockerfiles we use for building the images we ship to people... if I want to fix a bug with our Docker file that we use for releasing I might look at these and think that's what I need to fix.


Are they only used for GitHub actions? Can they go in the `.github` directory?

**Comment by dragonsahead** [comment]
> @paoliniluis , waiting on anything else to merge this?

I'm waiting till we release 42 as I don't want to push this and maybe cause an issue on our CI, just better safe than sorry

**Comment by dragonsahead** [comment]
> Can we put the new Dockerfiles inside a top-level directory? If I see a `java-8` and `java-17` at the top level I'm going to get confused and think there are source files in them, because we already have a top-level `java` directory with our one `.java` source file. If I look in them and see Docker files I'm maybe going to mistakenly think those are the Dockerfiles we use for building the images we ship to people... if I want to fix a bug with our Docker file that we use for releasing I might look at these and think that's what I need to fix.
> 
> Are they only used for GitHub actions? Can they go in the `.github` directory?

Done, I just moved everything into a single folder. These images will be used in our CircleCI and maybe also in GH actions, as they're now entirely controlled by us (and also ultra slim). The GH actions file is to build them now here rather than in dockerhub which is ultra slow and does not provide any feedback

**Review by camsaul — APPROVED** [review]

**Comment by pawit-metabase** [comment]
Can we get `docker` and `lein` as well (both available in `cimg/clojure`)?

 - `lein` is still used by Harbormaster as the build tool
 - `docker` is required as Harbormaster builds docker images during the build process

**Review by diogormendes — APPROVED** [review]

**Review by camsaul — CHANGES_REQUESTED** [review]
~We don't want to upgrade to Clojure 1.11.0 yet. We'll probably want to upgrade once they ship 1.11.1 or 1.11.0.1 or whatever the patch release ends up being called~

**Comment by dragonsahead** [comment]
> We don't want to upgrade to Clojure 1.11.0 yet. We'll probably want to upgrade once they ship 1.11.1 or 1.11.0.1 or whatever the patch release ends up being called

Thanks for the review @camsaul, I was trying to replicate our current clojure cli version in GH actions https://github.com/metabase/metabase/blob/master/.github/actions/prepare-backend/action.yml.

Nevertheless, I'm going to merge this when we ship 43, as I don't want to change anything when we're near the end of the release cycle. Once we go into 44, I'll merge this

**Review by camsaul — APPROVED** [review]
Is there not some way we could pass the Java version as a parameter to those Dockerfiles so we don't have to duplicate them between 8/11/17? Seems like a lot of repetition

Either way this is fine with me

**Comment by dragonsahead** [comment]
> Is there not some way we could pass the Java version as a parameter to those Dockerfiles so we don't have to duplicate them between 8/11/17? Seems like a lot of repetition
> 
> Either way this is fine with me

just did a test with build args and it works like an string interpolation, so yeah! 2 files only now 🚀

**Pull request merged**
